### DISCUSSION TITLE
Changes cause LTSP server to hang-up when client runs init-bottom

### DISCUSSION THREAD

**Discussion by bgpublic** [ltsp/ltsp#643] [Q&A]
I have been struggling for MANY, MANY hours trying to understand and fix an issue with fresh LTSP installations I've been trying to get going. (I've started from scratch twice and had the same issue occur both times, for seemingly different reasons).

I install fresh Lubuntu 18.04 32-bit (because I need 32-bit clients, and this seemed like the easiest way to do that).  I install Remote Desktop clients, video drivers from oibaf PPA, and I blacklist USB storage.  I make sure to completely update the OS.

I run an external DHCP server.   The server has a single NIC (it runs as a virtual machine on a Hyper-V host).  I have a permanent address set so that the DHCP server always assigns the LTSP server the address 192.168.50.6.  Everything else about that network (192.168.50.0/24) is pretty standard - internet and DNS work fine.   The NIC is dedicated to the LTSP server, and not shared with the host OS.

I follow the LTSP installation instructions exactly, including using the suggested PPA (add-apt-repository ppa:ltsp).

I do install pxelinux (apt-get pxelinux), because my old client hardware (HP T5740) need legacy boot.  On the DHCP server, I have the BootFile name point to the path/file for pxelinux.o (that I copied into /srv/tftp/ltsp/x86_32/).

The problem I am experiencing is that when the client system gets to /scripts/init-bottom, the server is frozen/hung-up at that point (caused by the client booting -- if I don't boot a client, then the server will not hang-up).  When I go to the Hyper-V server console, the KVM are completely hung-up for the LTSP server virtual machine.  See screenshot below with output FROM THE CLIENT's boot sequence.

My first fresh installation seemed to go fine until I tried to change the IP settings on the server.   I tried to disable IPv6 and the issue occurred.  I disabled IPv6 and the problem went away.   Next, though, I needed to move it to a different subnet, and that also caused the issue.   I went through A LOT of trial and error, research online, etc. and was not able to find any solution.   I gave up and decided to simply start with a fresh installation of Lubuntu.

The second fresh installation I made sure to get my IP settings finalized before even starting the LTSP installation.  I got everything else about my settings straight before doing the LTSP installation.  Everything worked fine, except that there were some OS updates that were held back by the system for some reason.  When the system finally applied the updates, the LTSP server started hanging-up on client boot, just like before.

I actually had one more effort this morning.   I had backed-up a copy of the LTSP server image from before LTSP was installed.  This time I made sure that all OS updates were completed before installing LTSP.   I'm experiencing the same issue now, without the system ever having booted successfully.

I'm guessing that something in the latest Ubuntu OS updates is causing the server to hang-up.  I do not know why in my initial effort the disabling of IPv6 would have caused the issue, and re-enabling IPv6 restored the system to boot properly again.  I figured that the system was simply really sensitive to IP configuration changes.   Please help.   I really like using LTSP, but need to upgrade from my outdated production instance. 

![image](https://user-images.githubusercontent.com/97898040/149797339-1c8f752f-6185-43be-90b4-77a99bfbb828.png)

**Comment by bgpublic** [comment]
Also, here is the output from ltsp info:

![image](https://user-images.githubusercontent.com/97898040/149803323-41733a39-6549-4712-b5dc-07ff8615437f.png)

**Comment by alkisg** [comment]
Hi @bgpublic, some quick notes:

1. You don't need pxelinux.0 for legacy clients. LTSP configures iPXE for UEFI, BIOS (Legacy) and Raspberry Pis, they can all work at the same time with the stock configuration.
2. Since you're experiencing hangs, I would avoid using the oibaf PPA and the USB blacklisting. Start with a clean updated Lubuntu 18.04 chrootless, then start customizing.
3. That said, it sounds to me like the hang is happening when the NFS server of the LTSP server is being accessed. Try to access it from any other client, without involving LTSP or netbooting at all. Example command:
    ```
    sudo mount -t nfs -o vers=3,nolock 192.168.50.6:/srv/ltsp /mnt
    ```
    If this indeed hangs the LTSP server VM, then your Hyper-V has bugs...

**Reply by bgpublic** [reply]
Thank-you for your rapid reply!

I updated my DHCP configuration to remove the BootFile.   The client machine still seemed to get the right DHCP info and began the boot process.   So, hopefully that's the end of that.   Thanks for the advice re: pxelinux.0.   The client boot does still hang the server.

I successfully mounted NFS from a 20.04 machine using the exact command that you provided (except I think you had the server address incorrect, so I used 192.168.50.6 instead).  The LTSP server kept humming along even after its NFS share was mounted.

I can easily roll-back to a copy of the drive image before I installed the video PPA, USB blacklist, pxelinux.0, and LTSP.   I'll do that now, make sure the OS is completely up-to-date, and then I will install LTSP.  I'll post back here with the result.

I REALLY appreciate your help!

**Reply by bgpublic** [reply]
Here is the result, @alkisg:

Even without video driver PPA and blacklist, the LTSP server still hangs upon client boot, just as before.

Tomorrow I am planning to roll back once again to an image before I've installed LTSP.  This time, I will not apply the latest updates and see if the problem still occurs once I install LTSP.   If it the problem does not occur, I'll save a copy of the image (so I can easily roll-back to when LTSP works) and then proceed with the updates to see if they trigger the issue somehow.

**Comment by alkisg** [comment]
Another idea is to dump the VM hard disk to a real PC and see if the problem happens on bare metal or not.
In general, the clients cannot make the server hang; it sounds like a bug in your hypervisor.

**Comment by bgpublic** [comment]
Hi, @alkisg:

I proceeded with a trial & error approach with the Ubuntu updates.  I can say with a pretty high degree of certainty that one/more of the updates in this screenshot are part of the issue that I am seeing.   I don't doubt that it could be the COMBINATION of one of these updates AND our Hyper-V installation, however I am pretty sure that the updates are indeed a trigger.

![image](https://user-images.githubusercontent.com/97898040/149995363-b7e0235f-ca3c-4beb-8ccf-924d2f0b36f3.png)

I should mention that our Hyper-V setup is a completely up-to-date Windows 10 machine running on Dell R710 hardware.

In order to tell whether Hyper-V is also a factor, I plan to install the same OS and LTSP on a Dell D630 laptop, and proceed to see whether these same updates break LTSP in that installation.

Please let me know if you have any other questions or insights.

Thank-you.

**Comment by bgpublic** [comment]
OK, Hyper-V definitely is a factor.   I cloned the working LTSP server from the virtual machine to the Dell D630 laptop.  I then completed all OS upgrades, updated LTSP image/initrd, and rebooted the server.   Clients boot fine and the server does not hang.

I'm still inclined to run my production LTSP server as a virtual machine and simply not apply these updates for the time being.   I'm hoping that whatever is wrong will be resolved some day and I'll be able to fully update at that time.   If there anything wrong with running the LTSP server as a Hyper-V virtual machine?   My current LTSP server (based on 14.04) has run reliably that way for many, many years -- although I was not good about OS updates. 

I really appreciate your help, @alkisg!

**Comment by alkisg** [comment]
@bgpublic, as you saw this problem is not related to LTSP at all.
In any case, you may try this command:

```shell
sudo apt install linux-generic-hwe-18.04
```

This will install the 5.4 kernel series instead of your older 4.15 series. See if that newer kernel works around the issue.
Other than that, I've never used Hyper-V, so I can't comment any more; personally I prefer VirtualBox.

**Reply by bgpublic** [reply]
Yes, @alkisg, you were right all along about the hypervisor.

I did try the updated kernel, installed just as you suggested, and it did not work around the issue.   I will figure-out a workable path (that might include VirtualBox).   Thanks a bunch for helping me get to this point!
